lib/backend/kubernetes/kubernetes.go

```diff
@@ -277,46 +288,61 @@ func (b *Backend) readSecretData(ctx context.Context, key backend.Key) (*backend
 	}, nil
 }
 
-func (b *Backend) updateSecretContent(ctx context.Context, items ...backend.Item) (*backend.Lease, error) {
+func isConflict(err error) bool { return kubeerrors.IsConflict(trace.Unwrap(err)) }
+
+func (b *Backend) updateSecretContent(ctx context.Context, isCreate bool, item backend.Item) (*backend.Lease, error) {
 	// FIXME(tigrato):
 	// for now, the agent is the owner of the secret so it's safe to replace changes
 
-	secret, err := b.getSecret(ctx)
-	if trace.IsNotFound(err) {
-		secret = b.genSecretObject()
-	} else if err != nil {
-		return nil, trace.Wrap(err)
-	}
+	if err := retry.OnError(retry.DefaultRetry, isConflict, func() error {
+		// NOTE: We create the secret in the constructor, so we know it exists at this point.
+		secret, err := b.getSecret(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
 
-	if secret.Data == nil {
-		secret.Data = map[string][]byte{}
-	}
+		if secret.Data == nil {
+			secret.Data = map[string][]byte{}
+		}
 
-	updateDataMap(secret.Data, items...)
+		k := backendKeyToSecret(item.Key)
+		// If the isCreate flag is set, fail if the key already exists.
+		if _, exists := secret.Data[k]; isCreate && exists {
+			// NOTE: This will not trigger a retry.
+			return trace.AlreadyExists("key %q already exists in secret %s", item.Key, b.SecretName)
+		}
+		secret.Data[k] = item.Value
 
-	if err := b.upsertSecret(ctx, secret); err != nil {
+		if err := b.upsertSecret(ctx, secret); err != nil {
```

Comment:
directly call update instead of upsert. that guarantees no one will ever break the system